okcupyd/json_search.py

```diff
@@ -118,17 +167,539 @@ def __call__(self, response_dictionary):
+## :type question_count_min: int
+
+
+# This key is available on profile.looking_for, but no corresponding field has yet been observed
+# on the new json_search API: single: True (is the target only looking for single matches?)
+
+
+
+# Mandatory criteria
+
+
+class LastLoginFilter(search_filters.filter_class):
+    output_key = "last_login"
+    description = "How recently returned search results must have been online."
+    accepted_values = ('hour', 'day', 'today', 'week', 'month', 'year', 'decade')
+    types = "str or int (seconds)"
+    def transform(last_online):
+        return(helpers.format_last_online(last_online))
+
+
+class RadiusFilter(search_filters.filter_class):
+    output_key = "radius"
+    decide = search_filters.all_decider # permit radius=None
+    description = "The maximum distance from the specified location of returned search results."
+    types = "int or None"
+    def transform(radius):
+        return radius
+
+
+class NearnessFilter(search_filters.filter_class):
+    output_key = "located_anywhere"
+    decide = search_filters.all_decider # permit radius=None
+    description = "The maximum distance from the specified location of returned search results."
+    types = "int or None"
+    def transform(radius):
+        return 1 if radius is None else 0
+
+
+class MinAgeFilter(search_filters.filter_class):
+    output_key = "minimum_age"
+    description = "The minimum age of returned search results."
+    types = int
+    # TODO: also permit single key `ages=(age_min, age_max)`, as in looking_for.py?
+    def transform(age_min=18):
+        return age_min
+
+
+class MaxAgeFilter(search_filters.filter_class):
+    output_key = "maximum_age"
+    description = "The maximum age of returned search results."
+    types = int
+    # TODO: also permit single key `ages=(age_min, age_max)`, as in looking_for.py?
+    def transform(age_max=99):
+        return age_max
+
+
+def accumulate_tags(keys, magicmap):
+    acc = 0
+    for k in keys:
+        i = magicmap[k]
+        if i is not None and i >= 0:
+            acc += 2 ** i
+        elif i < 0 or k.lower() in ('all', 'everyone', 'everybody', 'any'):
+            acc = 0
+            break
+        else:
+            raise TypeError
+    return acc
+
+
+# TODO also exploit `gender`?
+class GenderFilter(search_filters.filter_class):
+    output_key = "gender_tags"
+    decide = search_filters.any_not_none_decider
+    acceptable_values = (magicnumbers.maps.gender_tags.keys(), # also 'all', 'everyone', 'everybody', 'any'
+                        magicnumbers.gentation_to_number.keys())
+    types = ("string or list of strings", str)
+    def transform(gender_sought, gentation):
+        if gentation is not None:
+            assert gender_sought is None
+            gentation = gentation.strip().lower()
+            gent = magicnumbers.gentation_to_number[gentation]
+            if gent & (1+4+16):
```

Comment:
OK, but not sure how to concisely summarize all this. This is an attempt to implement all the complex interdependencies in the output_keys of `gender_tags`, `orientation_tags`, `gentation`, `i_want`, and `they_want`. (I don't know whether the last two are really needed in the requests, or are merely cosmetic, but I erred on the side of matching the web browser's requests closely.) On the input side, I imagined some users might be invoking searches using the `gentation` key, as we did with the old API. I provided the `gender_sought` and `orientation_sought` keys, which correlate to the new browser interface, and honor their expanded range of options. There's also the `gender` key from the `html_search.py` `SearchFetchable`, which I also documented in the `json_search.py` `SearchFetchable`, but haven't yet implemented any code that honors it. (That's documented as being the `gender` of the user running the search, not the gender being sought.)